src/structure/model.jl

```diff
@@ -0,0 +1,1089 @@
+Base.done(model::Model, state) = state > length(model)
+Base.eltype(::Type{Model}) = Chain
+
+# Iterating over a Chain yields AbstractResidues
+Base.length(chain::Chain) = length(resids(chain))
+Base.start(::Chain) = 1
+Base.next(chain::Chain, state) = (chain[resids(chain)[state]], state + 1)
+Base.done(chain::Chain, state) = state > length(chain)
+Base.eltype(::Type{Chain}) = AbstractResidue
+
+# Iterating over a Residue yields AbstractAtoms
+Base.length(res::Residue) = length(atomnames(res))
+Base.start(::Residue) = 1
+Base.next(res::Residue, state) = (res[atomnames(res)[state]], state + 1)
+Base.done(res::Residue, state) = state > length(res)
+Base.eltype(::Type{Residue}) = AbstractAtom
+
+# Iterating over a DisorderedResidue yields AbstractAtoms
+# This is not necessarily intuitive, it may be expected to yield Residues
+# However this way iterating over an AbstractResidue always yields AbstractAtoms
+Base.length(disordered_res::DisorderedResidue) = length(atomnames(disordered_res))
+Base.start(::DisorderedResidue) = 1
+Base.next(disordered_res::DisorderedResidue, state) = (defaultresidue(disordered_res)[atomnames(disordered_res)[state]], state + 1)
+Base.done(disordered_res::DisorderedResidue, state) = state > length(disordered_res)
+Base.eltype(::Type{DisorderedResidue}) = AbstractAtom
+
+# Iterating over an Atom returns itself
+# This is not necessarily intuitive, it may be expected to not be an iterator
+# However this way iterating over an AbstractAtom always yields Atoms
+Base.length(::Atom) = 1
+Base.start(::Atom) = 1
+Base.next(atom::Atom, state) = (atom, state + 1)
+Base.done(::Atom, state) = state > 1
+Base.eltype(::Type{Atom}) = Atom
+
+# Iterating over a DisorderedAtom yields Atoms
+Base.length(disordered_atom::DisorderedAtom) = length(altlocids(disordered_atom))
+Base.start(::DisorderedAtom) = 1
+Base.next(disordered_atom::DisorderedAtom, state) = (disordered_atom[altlocids(disordered_atom)[state]], state + 1)
+Base.done(disordered_atom::DisorderedAtom, state) = state > length(disordered_atom)
+Base.eltype(::Type{DisorderedAtom}) = Atom
+
+
+"""
+Returns a copy of a `Vector` of atoms or residues with all elements that do not
+satisfy `selector_functions...` removed.
+"""
+function applyselectors{T <: Union{AbstractResidue, AbstractAtom}}(element_list::Vector{T}, selector_functions::Function...)
+    new_list = copy(element_list)
+    applyselectors!(new_list, selector_functions...)
+    return new_list
+end
+
+"Runs `applyselectors` in place."
+function applyselectors!{T <: Union{AbstractResidue, AbstractAtom}}(element_list::Vector{T}, selector_functions::Function...)
+    for selector_function in selector_functions
+        filter!(selector_function, element_list)
+    end
+end
+
+applyselectors{T <: Union{AbstractResidue, AbstractAtom}}(element_list::Vector{T}) = copy(element_list)
+applyselectors!{T <: Union{AbstractResidue, AbstractAtom}}(element_list::Vector{T}) = element_list
```

Comment:
I think these two methods can be automatically derived from the above definitions. The `selector_functions...` argument accepts zero arguments.